RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc

```diff
@@ -44,10 +43,104 @@ void HcalDeterministicFit::getLandauFrac(float tStart, float tEnd, float &sum) c
 }
 
 void HcalDeterministicFit::phase1Apply(const HBHEChannelInfo& channelData,
-                                       const HcalCalibrations& calibs,
-                                       float* reconstructedEnergy,
-                                       float* reconstructedTime) const
+				       float& reconstructedEnergy,
+				       float& reconstructedTime) const
 {
-    // IMPLEMENT THIS!!!
-}
 
+  std::vector<double> corrCharge;
+  std::vector<double> inputCharge;
+  std::vector<double> inputPedestal;
+  double gainCorr = 0;
+  double respCorr = 0;
+
+  for(unsigned int ip=0; ip<channelData.nSamples(); ip++){
+
+    double charge = channelData.tsRawCharge(ip);
+    double ped = channelData.tsPedestal(ip); // ped and gain are not function of the timeslices but of the det ?
+    double gain = channelData.tsGain(ip);
+
+    gainCorr = gain;
+    inputCharge.push_back(charge);
+    inputPedestal.push_back(ped);
+
+  }
+
+  fPedestalSubFxn_.calculate(inputCharge, inputPedestal, corrCharge);
+
+  const HcalDetId& cell = channelData.id();
+
+  double fpar0, fpar1, fpar2;
+  if(std::abs(cell.ieta())<HcalRegion[0]){
+    fpar0 = fpars[0];
+    fpar1 = fpars[1];
+    fpar2 = fpars[2];
+  }else if(std::abs(cell.ieta())==HcalRegion[0]||std::abs(cell.ieta())==HcalRegion[1]){
+    fpar0 = fpars[3];
+    fpar1 = fpars[4];
+    fpar2 = fpars[5];
+  }else{
+    fpar0 = fpars[6];
+    fpar1 = fpars[7];
+    fpar2 = fpars[8];
+  }
+
+  if (fTimeSlew==0)respCorr=1.0;
+  else if (fTimeSlew==1)respCorr=rCorr[0];
+  else if (fTimeSlew==2)respCorr=rCorr[1];
+  else if (fTimeSlew==3)respCorr=frespCorr;
+
+  float tsShift3=HcalTimeSlew::delay(inputCharge[3], fTimeSlew, fTimeSlewBias, fpar0, fpar1 ,fpar2);
+  float tsShift4=HcalTimeSlew::delay(inputCharge[4], fTimeSlew, fTimeSlewBias, fpar0, fpar1 ,fpar2);
+  float tsShift5=HcalTimeSlew::delay(inputCharge[5], fTimeSlew, fTimeSlewBias, fpar0, fpar1 ,fpar2);
+
+  float i3=0;
+  getLandauFrac(-tsShift3,-tsShift3+tsWidth,i3);
+  float n3=0;
+  getLandauFrac(-tsShift3+tsWidth,-tsShift3+tsWidth*2,n3);
+  float nn3=0;
+  getLandauFrac(-tsShift3+tsWidth*2,-tsShift3+tsWidth*3,nn3);
+
+  float i4=0;
+  getLandauFrac(-tsShift4,-tsShift4+tsWidth,i4);
+  float n4=0;
+  getLandauFrac(-tsShift4+tsWidth,-tsShift4+tsWidth*2,n4);
+
+  float i5=0;
+  getLandauFrac(-tsShift5,-tsShift5+tsWidth,i5);
+  float n5=0;
+  getLandauFrac(-tsShift5+tsWidth,-tsShift5+tsWidth*2,n5);
+
+  float ch3=corrCharge[3]/i3;
```

Comment:
There are lots of potential divisions by zero on lines 113-128. `getLandauFrac` explicitly sets its last parameter to 0 under certain circumstances. It would be best check each value before dividing by it to make sure there are no zeroes.